### PR TITLE
[CI] Add Build Archive CI script.

### PR DESCRIPTION
**Pull request by elliotcmorris**
Part of https://github.com/OpenAssetIO/OpenAssetIO/issues/1181

Add archive behaviour to merges to main, so OpenAssetIO is built on all of our runner platforms and archived in github.

Intent is to unlock much better CI behaviour for us, as we are spending a lot of time and effort building OpenAssetIO across dependent repos, as well as just being generally convenient to have.

This will publish an artifact to every PR run, as well as one for each commit to main. The idea being that dependent repos are by default targetted to the `main` branch with the download-artifact action.

### CODE REVIEW

File: .github/workflows/build-archive.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        cmake --build build --parallel --config RelWithDebInfo
+
+        cmake --install build --config RelWithDebInfo
+      env:
+        CMAKE_TOOLCHAIN_FILE: ${{ github.workspace }}/.conan/conan_paths.cmake
+
+    - name: Upload archive
+      uses: actions/upload-artifact@v3
+      with:
+        name: openassetio-${{ runner.os }}
```
Comment (elliotcmorris) [inline .github/workflows/build-archive.yml:69]: The requirements stated we wanted to get the OS version as well, ie `windows 2019`, whereas this is just `windows`. However, i couldn't figure out a way to get that info without integrating a custom action, which didn't seem worthwhile to me.

File: .github/workflows/build-archive.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        cmake -S . -B build --install-prefix ${{ github.workspace }}/dist
+        -DOPENASSETIO_ENABLE_C=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo
+
+        cmake --build build --parallel --config RelWithDebInfo
+
+        cmake --install build --config RelWithDebInfo
+      env:
+        CMAKE_TOOLCHAIN_FILE: ${{ github.workspace }}/.conan/conan_paths.cmake
+
+    - name: Upload archive
```
Comment (foundrytom) [inline .github/workflows/build-archive.yml:66]: [question]
Im a bit concerned that we're not testing this build, and then planning to re-using it extensively downstream. This build stage could get out of sync with the one we are testing. It is also incredibly helpful when published artifacts correlate to other checks, especially for debugging. Would it be safer to publish the one from the main `ctest`  runs vs a custom untested build? This would need us to split build-test so we can make sure the artifact is uploaded even when the tests fail.

File: .github/workflows/build-archive.yml (hunk 1 above)
Comment (foundrytom) [reply .github/workflows/build-archive.yml:69]: `${{ matrix.config.os }}` should hopefully get you that value

**Review by foundrytom — CHANGES_REQUESTED** [review]
This is going to be mega useful, thanks. Main concern is that we're not testing this build.

File: .github/workflows/build-archive.yml (hunk 1 above)
Comment (elliotcmorris) [reply .github/workflows/build-archive.yml:69]: :facepalm:

File: .github/workflows/build-archive.yml (hunk 2 above)
Comment (elliotcmorris) [reply .github/workflows/build-archive.yml:66]: Because the build commands/runner configurations are duplicated, so we could be building an artifact configured different to the artifact we're running tests on?

Yeah I can see that, although I also saw it as a positive, as we gain coverage by more properly being a consumable artifact, (ie, not have the tests built in at all.)

Comment (foundrytom) [reply .github/workflows/build-archive.yml:66]: Yeah, exactly that. Def with you on the downstream repos using it - but if they also use the one we test here then I think you end up with the maximum amount of coverage?

Comment (foundrytom) [reply .github/workflows/build-archive.yml:66]: > (ie, not have the tests built in at all.)

Yeah, that would be ideal, would `ctest` be clever enough to be able to reuse just the `dist` part?

Comment (foundrytom) [reply .github/workflows/build-archive.yml:66]: Though if we publish the artifact before running `ctest` is there much difference? Aside from the chance the tests are using something outside of `dist`?

File: .github/workflows/build-archive.yml (hunk 1 above)
Comment (elliotcmorris) [reply .github/workflows/build-archive.yml]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1187/commits/5b334247f552ce12ae566b9ac320ad003bf99316
